In [ ]:
import pandas as pd
import numpy as np

In [ ]:
df1 = pd.read_csv("FA-KES-Dataset.csv",encoding ='unicode_escape')
df1.head()

unit_id  ... labels
0  1914947530  ...      0
1  1914947532  ...      0
2  1914947533  ...      0
3  1914947534  ...      0
4  1914947535  ...      0

[5 rows x 7 columns]

In [ ]:
df1 = df1.drop(['unit_id', 'article_title', 'source', 'date', 'location'], axis=1)
df1.columns = ['news', 'label']
df1

news  label
0    Wed 05 Apr 2017 Syria attack symptoms consiste...      0
1    Fri 07 Apr 2017 at 0914 Homs governor says U.S...      0
2    Sun 16 Apr 2017 Death toll from Aleppo bomb at...      0
3    Wed 19 Apr 2017 Aleppo bomb blast kills six Sy...      0
4    Sun 10 Jul 2016 29 Syria Rebels Dead in Fighti...      0
..                                                 ...    ...
799  28-08-2016 Turkish Bombardment Kills 20 Civili...      1
800  17-08-2016 Martyrs as Terrorists Shell Aleppos...      1
801  03-08-2016 Chemical Attack Kills Five Syrians ...      0
802  01-08-2016 5 Killed as Russian Military Choppe...      1
803  April 6 2017 Syrian Army Kills 48 ISIL Terrori...      1

[804 rows x 2 columns]

In [ ]:
df2 = pd.read_csv("fakenews-dataset.csv",error_bad_lines=False, engine="python")
df2.head()

Skipping line 6067: unexpected end of data


Unnamed: 0     id  ...                                               news label
0           0   9653  ...  WASHINGTON  —   In Sonny Perdue’s telling, Geo...     0
1           1  10041  ...  HOUSTON  —   Venezuela had a plan. It was a ta...     0
2           2  19113  ...  Sunday on ABC’s “This Week,” while discussing ...     0
3           3   6868  ...  AUGUSTA, Me.  —   The beleaguered Republican g...     0
4           4   7596  ...    Finian Cunningham has written extensively on...     1

[5 rows x 6 columns]

In [ ]:
df2.columns = ['number', 'id', 'headline', 'written_by', 'news', 'label']
df2.head()

number     id  ...                                               news label
0       0   9653  ...  WASHINGTON  —   In Sonny Perdue’s telling, Geo...     0
1       1  10041  ...  HOUSTON  —   Venezuela had a plan. It was a ta...     0
2       2  19113  ...  Sunday on ABC’s “This Week,” while discussing ...     0
3       3   6868  ...  AUGUSTA, Me.  —   The beleaguered Republican g...     0
4       4   7596  ...    Finian Cunningham has written extensively on...     1

[5 rows x 6 columns]

In [ ]:
df2 = df2.drop(['number', 'id', 'headline', 'written_by'], axis = 1)
df2 = df2[:550]
df2.head()

news  label
0  WASHINGTON  —   In Sonny Perdue’s telling, Geo...      0
1  HOUSTON  —   Venezuela had a plan. It was a ta...      0
2  Sunday on ABC’s “This Week,” while discussing ...      0
3  AUGUSTA, Me.  —   The beleaguered Republican g...      0
4    Finian Cunningham has written extensively on...      1

In [ ]:
df2

news  label
0    WASHINGTON  —   In Sonny Perdue’s telling, Geo...      0
1    HOUSTON  —   Venezuela had a plan. It was a ta...      0
2    Sunday on ABC’s “This Week,” while discussing ...      0
3    AUGUSTA, Me.  —   The beleaguered Republican g...      0
4      Finian Cunningham has written extensively on...      1
..                                                 ...    ...
545  ERBIL, Iraq  —   Mosul’s residents are hoardin...      0
546  Leave a Reply Click here to get more info on f...      1
547  People in parts of South America and Africa we...      0
548  Не верьте больше в Санту, дети, он хуже всех в...      1
549  Sen. Ted Cruz ( ) commended   Donald Trump’s c...      0

[550 rows x 2 columns]

In [ ]:
df = pd.concat([df1, df2])
df

news  label
0    Wed 05 Apr 2017 Syria attack symptoms consiste...      0
1    Fri 07 Apr 2017 at 0914 Homs governor says U.S...      0
2    Sun 16 Apr 2017 Death toll from Aleppo bomb at...      0
3    Wed 19 Apr 2017 Aleppo bomb blast kills six Sy...      0
4    Sun 10 Jul 2016 29 Syria Rebels Dead in Fighti...      0
..                                                 ...    ...
545  ERBIL, Iraq  —   Mosul’s residents are hoardin...      0
546  Leave a Reply Click here to get more info on f...      1
547  People in parts of South America and Africa we...      0
548  Не верьте больше в Санту, дети, он хуже всех в...      1
549  Sen. Ted Cruz ( ) commended   Donald Trump’s c...      0

[1354 rows x 2 columns]

In [ ]:
# Count how many hoax and valid label
df['label'].value_counts()

1    699
0    655
Name: label, dtype: int64

**Preprocessing Dataset**

In [ ]:
# Remove Punctuation

import string

def remove_punct(text):
  removed = "".join([i for i in text if i not in string.punctuation])
  return removed

In [ ]:
df['news'] = df['news'].apply(lambda x:remove_punct(x))
df.head()

news  label
0  Wed 05 Apr 2017 Syria attack symptoms consiste...      0
1  Fri 07 Apr 2017 at 0914 Homs governor says US ...      0
2  Sun 16 Apr 2017 Death toll from Aleppo bomb at...      0
3  Wed 19 Apr 2017 Aleppo bomb blast kills six Sy...      0
4  Sun 10 Jul 2016 29 Syria Rebels Dead in Fighti...      0

In [ ]:
# Lowercase All Words in Sentences

df['news'] = df.news.str.lower()
df

news  label
0    wed 05 apr 2017 syria attack symptoms consiste...      0
1    fri 07 apr 2017 at 0914 homs governor says us ...      0
2    sun 16 apr 2017 death toll from aleppo bomb at...      0
3    wed 19 apr 2017 aleppo bomb blast kills six sy...      0
4    sun 10 jul 2016 29 syria rebels dead in fighti...      0
..                                                 ...    ...
545  erbil iraq  —   mosul’s residents are hoarding...      0
546  leave a reply click here to get more info on f...      1
547  people in parts of south america and africa we...      0
548  не верьте больше в санту дети он хуже всех вра...      1
549  sen ted cruz   commended   donald trump’s choi...      0

[1354 rows x 2 columns]

In [ ]:
print(df['news'].iloc[0])

wed 05 apr 2017 syria attack symptoms consistent with nerve agent use who victims of a suspected chemical attack in syria appeared to show symptoms consistent with reaction to a nerve agent the world health organization said on wednesday some cases appear to show additional signs consistent with exposure to organophosphorus chemicals a category of chemicals that includes nerve agents who said in a statement putting the death toll at at least 70 the united states has said the deaths were caused by sarin nerve gas dropped by syrian aircraft russia has said it believes poison gas had leaked from a rebel chemical weapons depot struck by syrian bombs sarin is an organophosporus compound and a nerve agent chlorine and mustard gas which are also believed to have been used in the past in syria are not a russian defence ministry spokesman did not say what agent was used in the attack but said the rebels had used the same chemical weapons in aleppo last year the who said it was likely that some 

In [ ]:
# Train Test Split

from sklearn.model_selection import train_test_split

berita = df['news'].to_numpy()
y = df['label'].to_numpy()

x_train, x_test, y_train, y_test = train_test_split(berita, y, test_size = 0.2) # 80:20

In [ ]:
# Tokenize Data (Sentence Into Words)

from keras.preprocessing.text import Tokenizer
from keras.preprocessing.sequence import pad_sequences

tokenizer = Tokenizer(num_words=1000)
tokenizer.fit_on_texts(x_train)
tokenizer.fit_on_texts(x_test)
xsequences = tokenizer.texts_to_sequences(x_train)
xssequences = tokenizer.texts_to_sequences(x_test)

x_train_new = pad_sequences(xsequences)
x_test_new = pad_sequences(xssequences)


**Build Model**

In [ ]:
import tensorflow as tf

model = tf.keras.Sequential()
model.add(tf.keras.layers.Embedding(1000, 128))
# Using 1 LSTM Layer and 3 Dense Layer
model.add(tf.keras.layers.LSTM(128, dropout=0.4))
model.add(tf.keras.layers.Dense(496, activation='relu'))
model.add(tf.keras.layers.Dropout(0.2))
model.add(tf.keras.layers.Dense(215, activation='relu'))
model.add(tf.keras.layers.Dropout(0.5))
model.add(tf.keras.layers.Dense(1, activation='relu'))


model.compile(loss='binary_crossentropy', optimizer = 'adam', metrics=['accuracy'])

In [ ]:
# Callback Function

class callback(tf.keras.callbacks.Callback):
  def stop_epoch(self, epoch, logs={}):
    if(logs.get('accuracy')>0.96): # End iteration if accuracy > 96%
      print("End of Epochs")
      self.model.stop_epoch = True

callbacks = callback()

**Training Session**

In [ ]:
model.fit(x_train_new, y_train, batch_size=128, epochs=10,
          validation_data=(x_test_new, y_test),
          verbose=1, callbacks=[callbacks])

Epoch 1/10
9/9 [==============================] - 436s 48s/step - loss: 0.2215 - accuracy: 0.9185 - val_loss: 1.3916 - val_accuracy: 0.7778
Epoch 2/10
9/9 [==============================] - 436s 49s/step - loss: 0.2181 - accuracy: 0.9338 - val_loss: 1.4749 - val_accuracy: 0.7778
Epoch 3/10
9/9 [==============================] - 548s 59s/step - loss: 0.1951 - accuracy: 0.9329 - val_loss: 1.5832 - val_accuracy: 0.7969
Epoch 4/10
9/9 [==============================] - 328s 32s/step - loss: 0.1786 - accuracy: 0.9377 - val_loss: 1.6039 - val_accuracy: 0.8008
Epoch 5/10
9/9 [==============================] - 375s 39s/step - loss: 0.1695 - accuracy: 0.9473 - val_loss: 1.4595 - val_accuracy: 0.7893
Epoch 6/10
9/9 [==============================] - 390s 39s/step - loss: 0.1492 - accuracy: 0.9597 - val_loss: 1.7741 - val_accuracy: 0.8161
Epoch 7/10
9/9 [==============================] - 418s 43s/step - loss: 0.1793 - accuracy: 0.9569 - val_loss: 1.7394 - val_accuracy: 0.8046
Epoch 8/10
9/9 [====

**Testing Session**

In [ ]:
prediction = model.predict(x_test_new)

In [ ]:
# history = model.fit(x_test_new, y_test)

7/7 [==============================] - 61s 9s/step - loss: 0.9795 - accuracy: 0.8392


In [ ]:
history = model.fit(x_test_new, y_test)

9/9 [==============================] - 213s 24s/step - loss: 1.7493 - accuracy: 0.7816


-- It can be concluded that the model has the latest accuracy on the testing data of 78%